shell/gpu/gpu_surface_vulkan.cc

```diff
@@ -3,9 +3,47 @@
 // found in the LICENSE file.
 
 #include "flutter/shell/gpu/gpu_surface_vulkan.h"
+#include "lib/ftl/logging.h"
 
 namespace shell {
 
-//
+GPUSurfaceVulkan::GPUSurfaceVulkan(
+    ftl::RefPtr<vulkan::VulkanProcTable> proc_table,
+    std::unique_ptr<vulkan::VulkanNativeSurface> native_surface)
+    : window_(std::move(proc_table), std::move(native_surface)),
+      weak_factory_(this) {}
+
+GPUSurfaceVulkan::~GPUSurfaceVulkan() = default;
+
+bool GPUSurfaceVulkan::Setup() {
+  // This backend does not have an explicit setup task post initialization.
+  return window_.IsValid();
+}
+
+bool GPUSurfaceVulkan::IsValid() {
+  return window_.IsValid();
+}
+
+std::unique_ptr<SurfaceFrame> GPUSurfaceVulkan::AcquireFrame(
+    const SkISize& size) {
+  auto surface = window_.AcquireSurface();
+
+  if (surface == nullptr) {
+    return nullptr;
+  }
+
+  return std::make_unique<SurfaceFrame>(
+      std::move(surface),
+      [weak_this = weak_factory_.GetWeakPtr()](SkCanvas * canvas)->bool {
+        if (canvas == nullptr || !weak_this) {
```

Comment:
I'm not familiar with the semantics of whatever kind of weak pointer this is, but the one I'm familiar with guarantee safety by forcing you to obtain a temporary strong reference when you want to use it.  

This smells like a race condition waiting to happen.  If it cannot, there should be a comment about the program-wide invariant that guarantees this.

A similar comment applies to GPUSurfaceGL::AcquireFrame().